packages/cli/src/WorkflowExecuteAdditionalData.ts

```diff
@@ -664,16 +664,39 @@ function hookFunctionsSaveWorker(): IWorkflowExecuteHooks {
 						}
 					}
 
+					// Check config to know if execution should be saved or not
+					let saveDataErrorExecution = config.getEnv('executions.saveDataOnError') as string;
+					let saveDataSuccessExecution = config.getEnv('executions.saveDataOnSuccess') as string;
+					if (this.workflowData.settings !== undefined) {
+						saveDataErrorExecution =
+							(this.workflowData.settings.saveDataErrorExecution as string) ||
+							saveDataErrorExecution;
+						saveDataSuccessExecution =
+							(this.workflowData.settings.saveDataSuccessExecution as string) ||
+							saveDataSuccessExecution;
+					}
+
 					const workflowStatusFinal = determineFinalExecutionStatus(fullRunData);
 
-					if (workflowStatusFinal !== 'success') {
-						executeErrorWorkflow(
-							this.workflowData,
-							fullRunData,
-							this.mode,
-							this.executionId,
-							this.retryOf,
-						);
+					if (
+						(workflowStatusFinal === 'success' && saveDataSuccessExecution === 'none') ||
+						(workflowStatusFinal !== 'success' && saveDataErrorExecution === 'none')
+					) {
+						if (!fullRunData.waitTill) {
+							executeErrorWorkflow(
+								this.workflowData,
+								fullRunData,
+								this.mode,
+								this.executionId,
+								this.retryOf,
+							);
+							await Container.get(ExecutionRepository).hardDelete({
+								workflowId: this.workflowData.id as string,
+								executionId: this.executionId,
+							});
```

Comment:
Unfortunately this deletion cannot happen here.

Due to the nature of queue mode, we save the execution data to DB and notify via Redis that a given execution has finished.

This in turn allows the caller (be it `main` or `webhook`) to load the execution from DB (if needed) and react (such as sending an HTTP response or removing an item from a queue).

This code would need to be part of the array returned by `getWorkflowHooksWorkerMain` instead.